### ISSUE TITLE
What are the possible outcomes of interleaving SSL_write_early_data() and SSL_read()?

### ISSUE THREAD

**Issue by zhihan** [openssl/openssl#16486]
Hi, all 
In the documentation for SSL_write_early_data() (https://www.openssl.org/docs/man1.1.1/man3/SSL_write_early_data.html), it is mentioned that
"Once the initial SSL_write_early_data() call has completed successfully the client may interleave calls to SSL_read_ex(3) and SSL_read(3) with calls to SSL_write_early_data() as required."

I wonder what would happen in real applications. My understanding is:
(1) if the handshake is not over, SSL_read() will return SSL_ERROR_WANT_READ, and then we can continue to call SSL_write_early_data() to send early data.
(2) if the handshake is over, SSL_read() will return normally. But SSL_write_early_data() will error out with ERR_R_SHOULD_NOT_HAVE_BEEN_CALLED. 

Do I understand this correctly? Is it possible that both SSL_read() and SSL_write_early_data() succeed?

**Comment by zhihan** [comment]
It looks this [test](https://github.com/openssl/openssl/blob/e0f69c3598b61c47fbfe9d4e7d44ed671c334ef5/test/sslapitest.c#L3256) is testing a different behavior where server sends early data using `SSL_write_early_data()` and client reads the early data using `SSL_read_ex()`.

**Comment by mattcaswell** [comment]
Good to chat earlier.

I went back to the code and reminded myself how this works.

> Is it possible that both SSL_read() and SSL_write_early_data() succeed?

Yes!

There are 2 points in the handshake where we can pause to send early data:

1) Immediately after sending the ClientHello but before we've read anything from the server yet

2) Immediately before we send the client's certificate/cert verify/finished

Calling SSL_read() after SSL_write_early_data() will move the handshake forwards from state (1) to state (2). At this point the client has read and processed the server's flight and has authenticated the server. So it is reading "regular" data. It is still possible for it to send early data though because it hasn't yet completed sending its final flight.

Whether this is useful is another question. I don't know if anyone is actually doing this. But it is possible.

**Comment by zhihan** [comment]
Thanks, Matt.

**Comment by kaduk** [comment]
> Whether this is useful is another question. I don't know if anyone is actually doing this. But it is possible.

FWIW, the expectation of the TLS WG was that it is not useful to do this and that people would not do it.

**Comment by zhihan** [comment]
Thanks, Matt. For my understanding, if we are in state (2), would subsequent call to `SSL_read()` move the state on? If not, what function would move the state out of (2)?

**Comment by davidben** [comment]
> FWIW, the expectation of the TLS WG was that it is not useful to do this and that people would not do it.

I may be misinterpreting "this", but that doesn't match my understanding. After the ClientHello is sent, the client may stream early data before it receives ServerHello, or anywhere in between ServerHello and Finished. See this section in RFC8446: https://datatracker.ietf.org/doc/html/rfc8446#section-4.2.11.3

This is necessary when using 0-RTT with application protocols where either side may continuously write data. Notably HTTP/2. In an HTTP/2 client stack, calls into the network stack and bytes from the server may come in in any order. For instance, this is a possible interleaving:

1. The client is asked to fetch `GET /1`. It opens a new TCP connection to the server and sends ClientHello, followed by `GET /1` under 0-RTT.
2. The client receives ServerHello from the server, but not Finished.
3. The client is asked to fetch `GET /2`. (Maybe the user opened a new tab, or maybe there were a lot of images on this webpage.) It sends a new 0-RTT record with `GET /2`.
4. The client is asked to fetch `POST /3`. But this is not 0-RTT-able the handshake is not yet confirmed, so we don't send it yet.
5. The client is asked to fetch `GET /4`. It sends a new 0-RTT record with `GE`, but `T /4` did not fit in max_early_data_size.
6. The client receives EncryptedExtensions..Finished from the server. It sends EndOfEarlyData..Finished. The handshake is now complete.
7. The client finishes up `GET /4` by sending `T /4` over 1-RTT, and then it sends `POST /3`, which is no longer blocked.

**Comment by davidben** [comment]
An interesting consequence in HTTP, by the way, is that, in step (6), we might learn that _not only_ was 0-RTT rejected, but also the new connection negotiated `http/1.1` instead of `h2`. Then requests a-d can no longer proceed in parallel over the same connection. The natural way to implement this, IMO, to treat 0-RTT reject as a funny kind of error, then trigger a retry of each request. The HTTP stack's socket pooling logic can then rebind requests to sockets and/or spawn new ones as needed.

This is mostly outside of OpenSSL, but it means OpenSSL's API needs to be very clear on when connection properties can (0-RTT reject) and cannot (0-RTT accept) change. For BoringSSL, we ended up with a very different API, based on application needs for handling these two tricky cases.

**Comment by kaduk** [comment]
> I may be misinterpreting "this",

I mean, I could be, too.
But I thought Matt said that OpenSSL is letting the client send 0-RTT data after receiving the server Finished but before the client has sent its Certificate..Finished.  My recollection of the TLS WG is that "once you have the server Finished, don't send more with 0-RTT keys".

**Comment by davidben** [comment]
> My recollection of the TLS WG is that "once you have the server Finished, don't send more with 0-RTT keys".

That's right. Once you have server Finished, there's no reason not to just queue up a EndOfEarlyData + key_change + Finished for your next write.

However, there is a continuous interval of time in between sending the ClientHello and receiving server Finished, during which OpenSSL needs to _concurrently_ allow sending more early data and trying to processing ServerHello..Finished.

**Comment by kaduk** [comment]
> > My recollection of the TLS WG is that "once you have the server Finished, don't send more with 0-RTT keys".
> 
> That's right. Once you have server Finished, there's no reason not to just queue up a EndOfEarlyData + key_change + Finished for your next write.

Exactly

> However, there is a continuous interval of time in between sending the ClientHello and receiving server Finished, during which OpenSSL needs to _concurrently_ allow sending more early data and processing bytes of ServerHello..Finished as they come in.

Agreed.  Though OpenSSL's threading model serializes `SSL` access, so in practice the application's ability for concurrency may be quite limited.

**Comment by davidben** [comment]
Right, hence "concurrent" rather than "parallel". :-) The actual calls are serialized, but with non-blocking sockets, the high-level I/O operations are concurrent.

**Comment by zhihan** [comment]
> This is necessary when using 0-RTT with application protocols where either side may continuously write data. Notably HTTP/2. In an HTTP/2 client stack, calls into the network stack and bytes from the server may come in in any order. For instance, this is a possible interleaving:
> 
> 1. The client is asked to fetch `GET /a`. It opens a new TCP connection to the server and sends ClientHello, followed by `GET /a` under 0-RTT.
> 2. The client receives ServerHello from the server, but not Finished.
> 3. The client is asked to fetch `GET /b`. (Maybe the user opened a new tab, or maybe there were a lot of images on this webpage.) It sends a new 0-RTT record with `GET /b`.

Thanks for this example. Quick question: in this scenario, does the server send response to 'Get /a' and 'Get /b' to the client? And will client receive the responses before it writes end_of_early_data? I think in OpenSSL you can, which is part of my confusion originally.

**Comment by davidben** [comment]
> in this scenario, does the server send response to 'Get /a' and 'Get /b' to the client?

Yes. Getting the response is, well, sort of the point. :-)

> And will client receive the responses before it writes end_of_early_data? I think in OpenSSL you can, which is part of my confusion originally.

Depends a bit on how OpenSSL's implementation behaves, since writing a small amount of data is effectively synchronous. On the server, it will likely *send* the response before *receiving* EndOfEarlyData. Otherwise we've somewhat defeated the purpose of 0-RTT. This is the so-called "0.5-RTT" data.

However, 0.5-RTT and 1-RTT data is mostly indistinguishable on the client side, unless the client intentionally holds back Finished. A combined timeline for the basic single-request case looks something like:

1. Client sends ClientHello + 0-RTT request.
2. Transit time passes.
3. Server reads ClientHello. It sends ServerHello..Finished.
4. Server reads 0-RTT request. It sends response in 0.5-RTT, which are the same keys as 1-RTT.
5. Transit time passes.
6. Client receives ServerHello..Finished. It sends EndOfEarlyData/Finished and changes key.
7. Client receives response from server.

Whether the client decides to wait until it's sent EndOfEarlyData/Finished before progressing to step 7 is I guess an implementation detail, and mostly doesn't matter. (BoringSSL does wait because it makes our state machine a little simpler. But this is also all hidden behind our API, so the caller doesn't are.)

(Note these are timelines for what is _supposed to_ happen. I'm not familiar with OpenSSL's 0-RTT implementation and do not know whether it gets this right. I also don't know how this maps onto their APIs.)

**Comment by zhihan** [comment]
Whoops, sorry, I made edits after your reply.
> Yes. Getting the response is, well, sort of the point. :-)

I am thinking of a different case:
1. Client sends ClientHello + 0-RTT request.
2. Transit time passes.
3. Server reads ClientHello. Server reads 0-RTT request. It sends response. It sends ServerHello ... response in early_data ... Finished.
4. Transit time passes.
5. Client reads ServerHello .. response ..Finished. It sends EndOfEarlyData/Finished and changes key.

**Comment by kaduk** [comment]
> It sends response. It sends ServerHello ... response in early_data ... Finished.

This can't happen as described; the server's Finished is needed in order to produce the keys for protecting the 0.5-RTT data with the response.  But ServerHello..Finished, response in 0.5-RTT data can (and is expected to) happen.
Otherwise (with the corresponding modification to what the client reads in (5), your scenario looks like a valid case.

**Comment by zhihan** [comment]
Thanks for the response. If that case is not possible, what would be a possible scenario for the server to send early data back before `Finished`, as illustrated [in the standard](https://datatracker.ietf.org/doc/html/rfc8446#section-2.3)?

**Comment by kaduk** [comment]
I'm not sure which part of the section you're referring to, though I note that the string "early_data" that appears in figure 4 before the server's Finished refers to the extension named "early_data", sent in the EncryptedExtensions message.  The actual 0.5-RTT data is indicated by the `[Application Data*]`

**Comment by zhihan** [comment]
Ah, thanks for the clarification.

**Comment by mattcaswell** [comment]
> if we are in state (2), would subsequent call to SSL_read() move the state on? If not, what function would move the state out of (2)?

Calling SSL_read() at any point from state (1) until state (2) will move the handshake on - but it will not move you beyond state (2). Calling SSL_write() (as opposed to SSL_write_early_data()), or SSL_connect(), or SSL_do_handshake() would move you on to handshake completion.

> I may be misinterpreting "this", but that doesn't match my understanding. After the ClientHello is sent, the client may stream early data before it receives ServerHello, or anywhere in between ServerHello and Finished. 

Right! Exactly! So probably I wasn't clear in my original comment above. I believe this is the scenario that we're actually trying to cover here although OpenSSL supports sending early data all the way up until state (2) above which is technically the last point at which it is still valid to send early data (although not all that useful at that point). I believe the intention behind this design was to be able to send early data while in states (1) and (2) or *at any point between them*.

I have a slightly worry about whether this actually works. I note that we test writing early data in state (1) and in state (2), but we don't have a test for writing early data at some intermediate point between those two states. We probably should have.

**Comment by zhihan** [comment]
If my understanding is correct, this [test](https://github.com/openssl/openssl/blob/e0f69c3598b61c47fbfe9d4e7d44ed671c334ef5/test/sslapitest.c#L3283) is not possible. Particularly, when the server sends data (0.5-RTT data), the client must complete the handshake before it can read the data. Right?

**Comment by mattcaswell** [comment]
> If my understanding is correct, this test is not possible. Particularly, when the server sends data (0.5-RTT data), the client must complete the handshake before it can read the data. Right?

No. If the server sends 0.5RTT data the client can read it. It must have processed the server's handshake flight in order to do so, but it does not need to have sent its own final flight. As noted above this isn't particularly useful.

**Comment by zhihan** [comment]
> > If my understanding is correct, this test is not possible. Particularly, when the server sends data (0.5-RTT data), the client must complete the handshake before it can read the data. Right?
> 
> No. If the server sends 0.5RTT data the client can read it. It must have processed the server's handshake flight in order to do so, but it does not need to have sent its own final flight. As noted above this isn't particularly useful.

I see, so if the client choose to, it can continue to send early_data using the 0-RTT key. And the server can continue to read it and send response using the 1-RTT key. Does that summarize [this](https://github.com/openssl/openssl/blob/e0f69c3598b61c47fbfe9d4e7d44ed671c334ef5/test/sslapitest.c#L3293)?

**Comment by mattcaswell** [comment]
> I see, so if the client choose to, it can continue to send early_data using the 0-RTT key. And the server can continue to read it and send response using the 1-RTT key. Does that summarize this?

Correct.

**Comment by kaduk** [comment]
> > I see, so if the client choose to, it can continue to send early_data using the 0-RTT key. And the server can continue to read it and send response using the 1-RTT key. Does that summarize this?
> 
> Correct.

In terms of the crypto, that's correct, yes, but it would be a protocol violation to do so.
I didn't go and check the spec with particular care, but at least https://datatracker.ietf.org/doc/html/rfc8446#section-4.5 requires the client to send EndOfEarlyData after processing the server Finished, and once you send EndOfEarlyData you can't send more early data.

**Comment by mattcaswell** [comment]
> In terms of the crypto, that's correct, yes, but it would be a protocol violation to do so.

OpenSSL does not enforce this.

**Comment by t8m** [comment]
Closing as the issue was labeled as inactive for extended period of time and there was no additional input.
Please add a comment if you have any further information to support reopening the issue.
